### PR TITLE
Added support for finding all read groups when no read group specified

### PR DESCRIPTION
**Pull request by pkrusche**
Fixes issue #702

### CODE REVIEW

File: ga4gh/cli.py [hunk 1]
```diff
@@ -192,6 +192,18 @@ def getAllReadGroupSets(self):
...
 
+    def getAllReadGroups(self):
+        """
+        Get all read groups in a read group set
+        """
+        for dataset in self.getAllDatasets():
+            iterator = self._client.searchReadGroupSets(
+                datasetId=dataset.id)
+            for readGroupSet in iterator:
+                rgs = self._client.getReadGroupSet(readGroupSet.id)
```
Comment (dcolligan) [inline ga4gh/cli.py]: as a stylistic convention, we try and avoid abbreviations like this... use `readGroupSet` as a variable name instead

File: ga4gh/cli.py [hunk 2]
```diff
@@ -192,6 +192,18 @@ def getAllReadGroupSets(self):
...
+                for rg in rgs.readGroups:
```
Comment (dcolligan) [inline ga4gh/cli.py]: same style comment: `rg` -> `readGroup`

File: ga4gh/cli.py [hunk 3]
```diff
@@ -401,10 +434,10 @@ def _textOutput(self, gaObjects):
             # TODO add in some more useful output here.
             print(read.id)
 
-
 # ListReferenceBases is an oddball, and doesn't fit either get or
 # search patterns.
 
+
```
Comment (dcolligan) [inline ga4gh/cli.py]: the commit diffs should be as small as possible... try to avoid changing lines unnecessarily

File: ga4gh/cli.py [hunk 4]
```diff
@@ -385,13 +397,34 @@ def __init__(self, args):
         if args.readGroupIds is not None:
             self._readGroupIds = args.readGroupIds.split(",")
 
+    def _run(self, rgid, refid=None):
+        """
+        automatically guess reference id if not passed
+        """
+        # check if we can get reference id from rg
+        if not refid:
```
Comment (dcolligan) [inline ga4gh/cli.py]: checks for equivalence to None should read `if x is None`

File: ga4gh/cli.py [hunk 5]
```diff
@@ -385,13 +397,34 @@ def __init__(self, args):
...
+            refid = self._referenceId
+        if not refid:
```
Comment (dcolligan) [inline ga4gh/cli.py]: same comment here

File: tests/end_to_end/test_client_json.py [hunk 6]
```diff
@@ -73,3 +73,7 @@ def testGetDataset(self):
         for dataset in self._client.searchDatasets():
             self.verifyParsedOutputsEqual(
                 [dataset], "datasets-get", dataset.id)
+
+    def testSearchAllReadGroups(self):
+        cliOutput = self.captureJsonOutput("reads-search")
+        self.assertGreater(len(cliOutput), 0)
```
Comment (dcolligan) [inline tests/end_to_end/test_client_json.py]: there's probably a more rigorous test that we can write here... if the output is not what we expect, this test still passes

File: ga4gh/cli.py [hunk 7]
```diff
@@ -192,6 +192,18 @@ def getAllReadGroupSets(self):
...
+        """
+        Get all read groups in a read group set
+        """
+        for dataset in self.getAllDatasets():
+            iterator = self._client.searchReadGroupSets(
+                datasetId=dataset.id)
+            for readGroupSet in iterator:
+                rgs = self._client.getReadGroupSet(readGroupSet.id)
+                for rg in rgs.readGroups:
+                    yield rg.id
```
Comment (jeromekelleher) [inline ga4gh/cli.py]: We should yield the readgroup object here rather than the ID to keep symmetry with the other methods.

**Comment by dcolligan** [comment]
good start but still needs some work

**Comment by jeromekelleher** [comment]
Looks great @pkrusche, +1 modulo the stylistic nits above.

File: ga4gh/cli.py (hunk 3 above)
Comment (pkrusche) [inline ga4gh/cli.py]: I inserted this to make flake8 pass.

Comment (dcolligan) [inline ga4gh/cli.py]: hm, interesting... I wonder why this wasn't failing before, then...

make sure you're using the same version of flake that is specified in `requirements.txt`, if you aren't already...

Comment (pkrusche) [inline ga4gh/cli.py]: I installed via pip install -r requirements.txt, and used a virtualenv, so the version should be the same.

```
(server-ve)$:~/workspace/server$ flake8 --version
2.3.0 (pep8: 1.6.2, mccabe: 0.3.1, pyflakes: 1.0.0) CPython 2.7.9 on Linux
(server-ve)$:~/workspace/server$ cat requirements.txt | grep flake8
flake8==2.3.0  # Set due to conflict with pep8 in version 2.4.0
```

**Comment by dcolligan** [comment]
cool.  +1

**Pull request merged**